php/libraries/DefaultFilesUploadHandler.class.inc

```diff
@@ -0,0 +1,367 @@
+                                    "Could not move $oldPath to $newPath"
+                                );
+                            } else {
+                                $this->_movedFiles[$field] = $newFile;
+                            }
+                        }
+
+                        // Signal that an attempt to move a file from its temporary
+                        // location to its final location was made
+                        $this->onFileMoveEvent(
+                            $file['tmp_name'],
+                            isset($this->_errors[$field])
+                                ? null : $newFile->getRealPath(),
+                            $this->_errors[$field] ?? null
+                        );
+                    }
+                }
+            }
+        } finally {
+            $this->processingDone();
+        }
+    }
+
+    /**
+     * Finds the new (final, permanent) location for an uploaded file.
+     * This method only computes the new file path and does not actually
+     * move the file.
+     *
+     * @param string $tmpFilePath temporary path (on the server) where the
+     *               uploaded file is currently located.
+     *
+     * @return SplInfo the path information for the new file location.
+     */
+    public function getNewLocation(string $tmpFilePath) : SplFileInfo
+    {
+        foreach ($this->getFiles() as $formField => $file) {
+            if ($file['tmp_name'] == $tmpFilePath) {
+                // 1. Find the path that the file had on the client side
+                // 2. Take the basename and prepend $this->_moveDirPath
+                // 3. That's the new file path
+                $clientFile = new SplFileInfo($file['name']);
+                return new SplFileInfo(
+                    $this->_moveDirPath . '/' . $clientFile->getBaseName()
+                );
+            }
+        }
+
+        throw new InvalidArgumentException(
+            "Cannot find file with temporary upload path '$tmpFilePath'"
+        );
+    }
+
+    /**
+     * Method called when an attempt (either successful or unsuccessful) to move
+     * a file from its temporary location to its final location is made. By
+     * default this method does nothing but subclasses should override as needed
+     * (update a DB table, launch a computation, etc...)
+     *
+     * @param string $oldPath temporary path the file had on the server
+     *                        when it was uploaded by the client.
+     * @param string $newPath final path the file has on the server after it was
+     *                        moved to its final location (will be null if the file
+     *                        was not successfully moved)
+     * @param string $error   the error message describing why the file could not be
+     *                        moved (null if the file was moved successfully).
+     *
+     * @return nothing.
+     */
+    public function onFileMoveEvent(
```

Comment:
It might be a good idea to declare this as abstract rather than having a default which does nothing.

Reply:
I think providing a default no-op method makes sense (e.g for handlers that only store files in specific locations and do not do anything else)